specs/chugsplash-config.md

```diff
@@ -21,32 +21,90 @@ ChugSplash config files may be written in Javascript, Typescript, or JSON. For J
+- standard-transparent: Standard transparent proxies not deployed using OpenZeppelin.
+- oz-transparent: An OpenZeppelin transparent proxy
+- oz-ownable-uups: An OpenZeppelin UUPS proxy using Ownable
+- oz-access-control-uups: An OpenZeppelin UUPS proxy using Access Control
+
+## Previous Build Info
+A path to a build info file generated for a previous version of a contract that has already been deployed. This field needs to be defined when upgrading a contract that uses a proxy which was deployed outside of ChugSplash. Providing the previous build info allows us to run our storage slot checker
+
+## Previous Fully Qualified Name
+The fully qualified name of a contract that has been deployed outside of ChugSplash. This field needs to be defined along with the previous build info and is used to run the storage slot checker against the previous contract version.
+
+## Unsafe
+ChugSplash performs a variety of safety checks before deploying/upgrading contracts. Users may override these safety checks using this field if their situation requires it. The unsafe field is a JSON or JS object containing a set of fields which disable different safety checks including:
+- allowRenames: Allow renaming state variables
+- skipStorageCheck: Fully disable the storage slot checker
+- delegatecall: Allow use of delegatecall in implementation contracts
+- selfdestruct: Allow use of selfdestruct
+- missingPublicUpgradeTo: Allow upgrading to a contract that does not include an upgradeTo function (contracts using external UUPS proxies)
+- allowEmptyPush: Allow the use of the empty push function: `push()` which ChugSplash disallows due to potential ambiguity of the values in dynamic arrays if the function is used.
+
+# Contract Constructor Arguments
+We support providing a set of constructor arguements for each contract. We only support this because it is necessary to support the use of immutable variables. Therefore  all constructor arguements are required to be used by the constructor to set immutable variables in the target contract. If any logic is detected in the constructor that is not related to setting immutable variables than we thow an error. All other variables are expected to be set in the ChugSplash config file using the variable format defineed below.
+
+For constructor arguments we accept a JSON or Javascript object defining a set of constructor arguments. Since constructor arguments are expected to only be used to set immutable variables, we only support the following value types:
+- Signed and Unsigned integers
+- bytes32
+- boolean
+- addresses
+- User defined types
+
+For all of these types, we expect the same input format as the corresponding variable input format which is defined below.
+
+Note that for addresses, we support the use of contract references in constructor arguments as long as the target contract is proxied. We do not support contract references in the constructor arguements of non-proxied contracts.
```

Comment:
This is sort of separate to this PR, but I actually think the rule should be slightly different: contract references can only contain proxied contracts. In other words, `{{ MyNonProxiedContract }}` isn't allowed, but this is allowed:
```
MyNonProxiedContract: {
  constructorArgs: {
    {{ MyProxy }}
  }
}
```

I think it's fine to define the above config because we can resolve the contract references for proxies in advance without worrying about ordering.

Reply:
Yeah, I think that's a good change and we should update this spec to include it.

Reply:
Actually, after thinking about this more I don't think we should make this change. I'll post more on the relevant ticket.